examples/smart-contract-upgrade/Contract_Version2/src/lib.rs

```diff
@@ -0,0 +1,212 @@
+#[receive(
+    contract = "smart_contract_upgrade",
+    name = "migration",
+    error = "CostumContractError",
+    low_level
+)]
+fn contract_migration<S: HasStateApi>(
+    ctx: &impl HasReceiveContext,
+    host: &mut impl HasHost<S>,
+) -> ContractResult<()> {
+    // Check that only this contract instance can call this function.
+    ensure!(ctx.sender().matches_contract(&ctx.self_address()), CostumContractError::Unauthorized);
+
+    // Read the old top-level contract state.
+    let state: OldState = host.state().read_root()?;
+
+    let new_state = State {
+        admin:     state.admin,
+        old_state: state.to_be_migrated_state,
+        new_state: "This is the new state.".to_string(),
+    };
+
+    host.state_mut().write_root(&new_state);
+    host.commit_state();
+
+    Ok(())
+}
+
+/// Upgrade this smart contract instance to a new module and call optionally a
+/// migration function after the upgrade.
+///
+/// It rejects if:
+/// - Sender is not the admin of the contract instance.
+/// - It fails to parse the parameter.
+/// - If the ugrade fails.
+/// - If the migration invoke fails.
+///
+/// This function is marked as `low_level`. This is **necessary** since the
+/// high-level mutable functions store the state of the contract at the end of
+/// execution. This conflicts with migration since the shape of the state
+/// **might** be changed by the migration function. If the state is then written
+/// by this function it would overwrite the state stored by the migration
+/// function.
+#[receive(
+    contract = "smart_contract_upgrade",
+    name = "upgrade",
+    parameter = "UpgradeParams",
+    error = "CostumContractError",
+    low_level
+)]
+fn contract_upgrade<S: HasStateApi>(
+    ctx: &impl HasReceiveContext,
+    host: &mut impl HasHost<S>,
+) -> ContractResult<()> {
+    // Read the top-level contract state.
+    let state: State = host.state().read_root()?;
+
+    // Check that only the admin is authorized to upgrade the smart contract.
+    ensure!(ctx.sender().matches_account(&state.admin), CostumContractError::Unauthorized);
+    // Parse the parameter.
+    let params: UpgradeParams = ctx.parameter_cursor().get()?;
+    // Trigger the upgrade.
+    host.upgrade(params.module)?;
+    // Call the migration function if provided.
+    if let Some((func, parameters)) = params.migrate {
+        host.invoke_contract_raw(
+            &ctx.self_address(),
+            parameters.as_parameter(),
+            func.as_entrypoint_name(),
+            Amount::zero(),
+        )?;
+    }
+    Ok(())
+}
+
+/// Transfer the admin address to a new admin address.
+///
+/// It rejects if:
+/// - Sender is not the current admin of the contract instance.
+/// - It fails to parse the parameter.
+#[receive(
+    contract = "smart_contract_upgrade",
+    name = "updateAdmin",
+    parameter = "AccountAddress",
+    error = "CostumContractError",
+    mutable
+)]
+fn contract_update_admin<S: HasStateApi>(
```

Comment:
I'd probably even remove this function since it is not really needed for the example. And it would make it simpler.

Reply:
The admin is just set at initialization.